plonk/src/circuit/basic.rs

```diff
@@ -1416,6 +1488,105 @@ impl<F: PrimeField> PlonkCircuit<F> {
+        let a_le_const =
+            self.is_leq_constant_internal(a, &F::from(F::modulus_minus_one_div_two()))?;
+        let b_le_const =
+            self.is_leq_constant_internal(b, &F::from(F::modulus_minus_one_div_two()))?;
+        let b_ge_const = self.logic_neg(b_le_const)?;
+        // Check whether `a` <= (q-1)/2 and `b` > (q-1)/2
+        let msb_check = self.logic_and(a_le_const, b_ge_const)?;
+
+        // Check whether `a` and `b` are both <= (q-1)/2 or
+        // are both > (q-1)/2
+        let msb_eq = self.check_equal(a_le_const.0, b_le_const.0)?;
+        // If so, check whether (b-a) <= (q-1)/2
+        let c = self.sub(b, a)?;
+        let cmp_result =
+            self.is_leq_constant_internal(c, &F::from(F::modulus_minus_one_div_two()))?;
+        let cmp_result = self.logic_and(msb_eq, cmp_result)?;
+
+        self.logic_or_gate(msb_check, cmp_result)
+    }
+
+    /// Helper function to check whether `a` is no greater
+    /// than a given constant. Let N = F::size_in_bits()
+    /// It uses N negation gates and N AND/OR gates.
+    fn is_leq_constant_internal(
+        &mut self,
+        a: Variable,
+        constant: &F,
+    ) -> Result<BoolVar, PlonkError> {
+        let a_bits_le = self.unpack(a, F::size_in_bits())?;
+        let const_bits_le = constant.into_repr().to_bits_le();
+        let bits_length = const_bits_le.len().max(a_bits_le.len());
+
+        // Stack for constructing the comparison gate
+        let mut current_run = Vec::new();
+        // Iterate from LSB to MSB. If current bit of constant is 0
+        // push a NOT(a) to the stack. Otherwise, wrap and clear the
+        // current stack using an k-ary AND gate, OR the result with
+        // a NOT(a) and push it back to the stack.
+        const_bits_le
+            .into_iter()
+            .take(bits_length)
+            .zip(a_bits_le.iter().take(bits_length))
+            .try_for_each(|(b, a)| -> Result<(), PlonkError> {
+                if b {
+                    let c = self.logic_neg(*a)?;
+                    let d = self.logic_and_all(&current_run)?;
+                    let c = self.logic_or(c, d)?;
+                    current_run.clear();
+                    current_run.push(c);
+                } else {
+                    let c = self.logic_neg(*a)?;
+                    current_run.push(c);
+                }
+                Ok(())
+            })?;
+        // If bits length of `a` is too long, check whether the most significant
+        // bits of `a` is 0.
+        a_bits_le
+            .iter()
+            .skip(bits_length)
+            .try_for_each(|a| -> Result<(), PlonkError> {
+                let c = self.logic_neg(*a)?;
+                current_run.push(c);
+                Ok(())
+            })?;
+        if current_run.len() == 1 {
+            Ok(current_run[0])
+        } else {
+            self.logic_and_all(&current_run)
```

Comment:
Instead of using a stack, we only need to obtain 2 vars finally, `c` that as mentioned before, and `d` used to check whether `ak, a_{k+1}, ..., a_n` are all zeroes. Finally we return `d && NOT(c)`